File: src/main/java/net/pterodactylus/sone/utils/Optionals.java
package net.pterodactylus.sone.utils;

import com.google.common.base.Function;
import com.google.common.base.Optional;
import com.google.common.base.Predicate;

/**
 * Helper methods for dealing with {@link Optional}s.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class Optionals {

	public static Predicate<Optional<?>> isPresent() {
		return new Predicate<Optional<?>>() {
			@Override
			public boolean apply(Optional<?> input) {
				return input.isPresent();
			}
		};
	}

	public static <T> Function<Optional<T>, T> get() {
		return new Function<Optional<T>, T>() {
			@Override
			public T apply(Optional<T> input) {
				return input.get();
			}
		};
	}

}

File: src/main/kotlin/net/pterodactylus/sone/utils/Optionals.kt
package net.pterodactylus.sone.utils

import com.google.common.base.Optional

fun <T, R> Optional<T>.let(block: (T) -> R): R? = if (isPresent) block(get()) else null
fun <T> Optional<T>.also(block: (T) -> Unit): Optional<T> { if (isPresent) block(get()); return this }

fun <T> T?.asOptional(): Optional<T> = this?.let { Optional.of(it) } ?: Optional.absent<T>()

fun <T, R> Iterable<T>.mapPresent(transform: (T) -> Optional<R>): List<R> =
		map(transform).filter { it.isPresent }.map { it.get() }
